debian/chinit.1
.TH CHINIT 1


.SH NAME
chinit - change init (run process in new pid namespace)


.SH SYNOPSYS
.B chinit
[\fIoptions\fR]
[\fIcommand\fR]


.SH DESCRIPTION
.B chinit
executes \fIcommand\fR in new pid namespace;
see
.BR  pid_namespace (7)
for more details.
Closing of created pid namespace is garanted
in cases of termination of execututed \fIcommand\fR
or any of auxiliary processes,
including initially launched one.
It is guaranteed by kernel
that all descendant processes will be destroyed.


.SH OPTIONS
.TP
.BR \-e ", " \-\-exec
Perform exec to given \fIcommand\fR in 'init' process of new namespace instead of vfork/exec pair;
see
.BR NOTES
for details.
.TP
.BR \-p ", " \-\-proc
Remount
.BR proc (5).
.TP
.BR \-h ", " \-\-help
Show short help and exit.


.SH EXIT STATUS
Normally
.B chinit
returns exit status of executed command.
.br
If any of auxiliary processes is killed by a signal,
returns the appropriate status (128 + WTERMSIG(status)).
.br
If internal error occurred, returns EXIT_FAILURE.


.SH NOTES
.B chinit
implementation scheme:
    root process
        with \-\-proc:
            unshare(CLONE_NEWPID|CLONE_NEWNS) (slower!)
        else
            unshare(CLONE_NEWPID)
        fork -> waitpid for child process
        with \-\-proc:
            remount /proc/
        drop privileges
    child process ('init' for new pid namespace):
        set prctl(PR_SET_PDEATHSIG, SIGKILL)
        with --exec option:
            execvp(\fIcommand\fR) (unsafe for some commands)
        else:
            vfork -> waitpid for child process
            execvp(\fIcommand\fR)

According to
.BR prctl (2),
PR_SET_PDEATHSIG will be cleared:
  * for the child of a fork
  * when executing a set-user-ID or set-group-ID binary
.br
Since it is difficult to determine if the parent will reset this flag, then an additional fork used by default.
.br
If \fIcommand\fR does not do anything of the above, is is possible to use
.BR \-\-exec
flag and save one fork.


.SH BUGS
The best way to report a bug is to create an issue on https://github.com/romanoved/chinit/issues .


.SH AUTHOR
Evgenii D. Romanov <romanoved@yandex.ru>


.SH SEE ALSO
.BR prctl (2),
.BR unshare (2),
.BR vfork (2),
.BR waitpid (2),
.BR exec (3),
.BR proc (5),
.BR pid_namespace (7)


debian/changelog
chinit (1.4) stable; urgency=medium

  * Use --no-proc by default.
  * Remove -P/--no-proc.
  * Update man page.

 -- Evgenii Romanov <romanoved@yandex.ru>  Fri, 04 Sep 2020 13:01:44 +0300

chinit (1.3) stable; urgency=medium

  * Fix missed logic about remount procfs(v1.0) and vfork(v1.1).
  * Clang-format on source.
  * Add -p/--proc and -P/--no-proc options.

 -- Evgenii Romanov <romanoved@yandex.ru>  Wed, 22 Mar 2017 20:51:09 +0300

chinit (1.2) stable; urgency=medium

  * Fix Makefile.

 -- Evgenii Romanov <romanoved@yandex.ru>  Tue, 21 Mar 2017 23:09:02 +0300

chinit (1.1) unstable; urgency=medium

  * Add --exec option.
  * Fix exit status for abnormal exit.
  * Add man page.

 -- Evgenii Romanov <romanoved@yandex.ru>  Tue, 21 Mar 2017 17:19:04 +0300

chinit (1.0) unstable; urgency=medium

  * Remount procfs.

 -- Evgenii Romanov <romanoved@yandex.ru>  Wed, 24 Jun 2015 00:26:44 +0300

chinit (0.9) unstable; urgency=low

  * Improve implementation using prctl syscall instead of action on pipe.

 -- Evgenii Romanov <romanoved@yandex.ru>  Tue, 09 Jun 2015 18:52:14 +0300

chinit (0.8) unstable; urgency=low

  * Initial release.

 -- Evgenii Romanov <romanoved@yandex.ru>  Thu, 18 Dec 2014 01:53:29 +0300


chinit.c
#ifndef _GNU_SOURCE
#define _GNU_SOURCE
#endif

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>

#include <sys/mount.h>
#include <sys/prctl.h>
#include <sys/types.h>
#include <sys/wait.h>

#include <sched.h>
#include <unistd.h>

#include <getopt.h>

/* simple logging helpers */

static struct {
    const char* main_cmd;
    int is_proc;
    int is_exec;
    const char* sub_cmdline;
} logger_data;

char* get_cmdline(int argc, char* const argv[], int opt_index) {
    size_t total_size = 0;
    int i;
    char* cmdline;
    char* dest;

    for (i = opt_index; i < argc; ++i)
        total_size += strlen(argv[i]) + 1;

    if (!(cmdline = (char*)malloc(total_size))) {
        perror("get_cmdline");
        abort();
    }

    memset(cmdline, ' ', total_size);
    cmdline[total_size - 1] = '\0';

    dest = cmdline;
    for (i = opt_index; i < argc; ++i) {
        memcpy(dest, argv[i], strlen(argv[i]));
        dest += strlen(argv[i]) + 1;
    }
    return cmdline;
}

int log_exit_failure(const char* msg) {
    fprintf(stderr, "%s %s%s: %s (%s); cmd '%s'\n",
            logger_data.main_cmd,
            logger_data.is_proc ? "-p" : "-P",
            logger_data.is_exec ? "e" : "",
            msg, strerror(errno),
            logger_data.sub_cmdline);
    return EXIT_FAILURE;
}

void print_help(char* chinit_cmd) {
    fprintf(
        stderr,
        "Usage: %s [options] [command]\n\n"
        "Execute command in new temporary pid namespace.\n\n"
        "Options:\n"
        "    -h, --help: show this help and exit;\n"
        "    -e, --exec: do plain exec instead of fork/exec pair (see man page for details);\n"
        "    -p, --proc: remout /proc (default for now);\n"
        "    -P, --no-proc: do not remout /proc;\n\n"
        ""
        "See man page for details about --exec/--proc/--no-proc options.\n",
        chinit_cmd);
}

int main(int argc, char* const argv[]) {
    int is_exec = 0;
    int is_proc = 0;

    if (argc == 1) {
        print_help(*argv);
        return EXIT_FAILURE;
    }

    while (1) {
        const struct option long_options[] =
            {
                {"help", no_argument, 0, 'h'},
                {"exec", no_argument, 0, 'e'},
                {"proc", no_argument, 0, 'p'},
                {0, 0, 0, 0}};

        int getopt_rc;
        int option_index = 0;
        if ((getopt_rc = getopt_long(argc, argv, "+hep", long_options, &option_index)) == -1)
            break;
        switch (getopt_rc) {
            case 'h':
                print_help(*argv);
                return EXIT_FAILURE;
            case 'e':
                is_exec = 1;
                break;
            case 'p':
                is_proc = 1;
                break;
            case '?':
                /* getopt_long already printed an error message. */
                return EXIT_FAILURE;
            default:
                fprintf(stderr, "%s: unhandled return value of getopt_long: %d\n", argv[0], getopt_rc);
                return EXIT_FAILURE;
        }
    }

    /* prepare cmdline for pretty exception printing */
    logger_data.main_cmd = argv[0];
    logger_data.is_proc = is_proc;
    logger_data.is_exec = is_exec;
    logger_data.sub_cmdline = get_cmdline(argc, argv, optind);

    /* arguments of sub_command*/
    argv += optind;

    /* fork of current process after unshare */
    /* will be an analogue of the init for new pid namespace */

    if (is_proc) {
        if (unshare(CLONE_NEWPID | CLONE_NEWNS) < 0)
            return log_exit_failure("unshare(CLONE_NEWPID|CLONE_NEWNS) failed");
    } else {
        if (unshare(CLONE_NEWPID) < 0)
            return log_exit_failure("unshare(CLONE_NEWPID) failed");
    }

    {
        pid_t pid;
        int status = 0;

        if ((pid = fork()) < (pid_t)0)
            return log_exit_failure("first fork failed");

        if (pid != (pid_t)0) {
            /* parent process waits for the completion of a child */
            if (waitpid(pid, &status, 0) < 0)
                return log_exit_failure("first fork: waitpid failed");
            return WIFEXITED(status) ? WEXITSTATUS(status) : 128 + WTERMSIG(status);
        }
    }

    /* chid after fork - init of new pid namespace */

    /* remount proc */

    if (is_proc) {
        if (mount("none", "/proc", NULL, MS_PRIVATE | MS_REC, NULL))
            return log_exit_failure("failed to unmount old /proc");

        if (mount("proc", "/proc", "proc", MS_NOSUID | MS_NOEXEC | MS_NODEV, NULL))
            return log_exit_failure("failed to mount new /proc");
    }

    /* drop_privileges */
    {
        uid_t userid = 0;
        gid_t groupid = 0;

        groupid = getgid();
        if (setgid(groupid) != 0)
            return log_exit_failure("dropping privileges: setgid failed");

        userid = getuid();
        if (setuid(userid) != 0)
            return log_exit_failure("dropping privileges: setuid failed");
    }

    /* expect SIGKILL in child if parent process fail */

    if (prctl(PR_SET_PDEATHSIG, SIGKILL))
        return log_exit_failure("prctl failed");

    /* without --exec do additional fork */
    if (!is_exec) {
        pid_t pid;
        int status = 0;

        if ((pid = vfork()) < (pid_t)0)
            return log_exit_failure("second vfork failed");

        if (pid != (pid_t)0) {
            if (waitpid(pid, &status, 0) < 0)
                return log_exit_failure("first fork: waitpid failed");
            return WIFEXITED(status) ? WEXITSTATUS(status) : 128 + WTERMSIG(status);
        }
    }

    execvp(*argv, argv);

    /* this code should never be reached if execvp successful */
    _exit(log_exit_failure("execvp failed"));
}
